### PR TITLE
MAST: Moving target TESScut

### PR DESCRIPTION
**Pull request by ceb8**
Adding functionality to access the new MAST Moving Target TESScut functionality from the astroquery.mast Tesscut class.

Includes:

- New `moving_target` argument to Tesscut class functions, that allows users to make cutouts for named moving targets (asteroids/comets etc).
- Associated documentation
- Associated tests (remote and local)
- Update to mast cloud access documentation to reflect prior code changes (unrelated, just apparently forgot to do the docs when I updated the code)
- General test tweaks to explicitly catch warnings and make sure all the remote tests pass

### CODE REVIEW

File: astroquery/mast/cutouts.py [hunk 1]
```diff
@@ -117,36 +120,64 @@ def get_sectors(self, coordinates=None, radius=0*u.deg, objectname=None):
...
+            # Add optional parameter is present
+            if mt_type:
+                params["obj_type"] = mt_type
+
+            response = self._service_api_connection.service_request_async("mt_sector", params)
+
+        else:
+
+            # Get Skycoord object for coordinates/object
+            coordinates = parse_input_location(coordinates, objectname)
```
Comment (bsipocz) [inline astroquery/mast/cutouts.py:168]: I see that this has been the same before, but maybe it's worth using a different name here and below in the other methods, too that `coordinates` that is also a non-parsed, and maybe None input kwarg.

File: astroquery/mast/cutouts.py [hunk 2]
```diff
@@ -117,36 +120,64 @@ def get_sectors(self, coordinates=None, radius=0*u.deg, objectname=None):
...
             The string must be parsable by `~astropy.coordinates.Angle`. The
             appropriate `~astropy.units.Quantity` object from
             `astropy.units` may also be used.
+            This argument is ignored if moving_target is supplied.
         objectname : str, optional
             The target around which to search, by name (objectname="M104")
             or TIC ID (objectname="TIC 141914082").
-            One and only one of coordinates and objectname must be supplied.
+            One and only one of coordinates, objectname, moving_target must be supplied.
+        moving_target : str, optional
```
Comment (bsipocz) [inline astroquery/mast/cutouts.py:135]: Why do we need a new kwarg for moving targets? At the end of the day they are the name of the object to query, and from the user's pow I don't see a difference whether querying for M1 or Ceres.

File: astroquery/mast/observations.py [hunk 3]
```diff
@@ -546,11 +546,18 @@ def download_file(self, uri, local_path=None, base_url=None, cache=True, cloud_o
         url = None
 
         try:
-            if self._cloud_connection is not None and self._cloud_connection.is_supported(data_product):
-                try:
-                    self._cloud_connection.download_file(data_product, local_path, cache)
-                except Exception as ex:
-                    log.exception("Error pulling from S3 bucket: {}".format(ex))
+            if self._cloud_connection is not None:
```
Comment (bsipocz) [inline astroquery/mast/observations.py:549]: how is this related to this PR?

File: docs/mast/mast.rst [hunk 4]
```diff
@@ -324,35 +286,77 @@ Product filtering can also be applied directly to a table of products without pr
...
+
+Downloading Data Products
+-------------------------
+
+Products can be downloaded by using `~astroquery.mast.ObservationsClass.download_products`,
+with a `~astropy.table.Table` of data products, or a list (or single) obsid as the argument.
+
+.. code-block:: python
+
+                >>> from astroquery.mast import Observations
```
Comment (bsipocz) [inline docs/mast/mast.rst:298]: this looks overindented

File: docs/mast/mast.rst [hunk 5]
```diff
@@ -324,35 +286,77 @@ Product filtering can also be applied directly to a table of products without pr
...
+​As an alternative to downloading the data files now, the ``curl_flag`` can be used instead to instead get a curl script that can be used to download the files at a later time.
+
+.. code-block:: python
+
+                >>> from astroquery.mast import Observations
+
+                >>> single_obs = Observations.query_criteria(obs_collection="IUE",obs_id="lwp13058")
+                >>> data_products = Observations.get_product_list(single_obs)
+                
+                >>> Observations.download_products(data_products,
```
Comment (bsipocz) [inline docs/mast/mast.rst:323]: `download_products` returns a table, maybe it's worth assigning it into a variable. However, it stays as is (I really don't have a strong preference), then include the table in the output below.

File: docs/mast/mast.rst [hunk 6]
```diff
@@ -813,13 +808,35 @@ ccd a target pixel file will be produced for each one.
...
+Requesting a cutout by moving_target accesses the
+`MAST Moving Target TESScut API <https://mast.stsci.edu/tesscut/docs/getting_started.html#moving-target-cutouts>`__
+and returns a target pixel file, with format described
+`here <https://astrocut.readthedocs.io/en/latest/astrocut/file_formats.html#path-focused-target-pixel-files>`__.
+The moving_target may be any object name or ID understood by the
+`JPL Horizonf ephemerades interface <https://ssd.jpl.nasa.gov/horizons.cgi>`__.
+
+.. code-block:: python
+
+                >>> from astroquery.mast import Tesscut
```
Comment (bsipocz) [inline docs/mast/mast.rst:820]: I run into a timeout with this example

**Review by bsipocz — COMMENTED** [review]
This certainly needs a changelog entry. 

Most of my comments are nit picks, but while I see why it's easier for the backend to use a new kwargs, for UX I don't think we need a new kwarg to specify the name.


Some of the test/cloud  changes are unrelated to this PR, now it's fine to leave them, but next time I think they worth to put them in separate, smaller PRs.


I also have one remote failure:

```
astroquery/mast/tests/test_mast.py .........................................                                                                  [ 51%]
astroquery/mast/tests/test_mast_remote.py .........................F...........                                                               [ 98%]
docs/mast/mast.rst s                                                                                                                          [100%]

===================================================================== FAILURES ======================================================================
_______________________________________________________ TestMast.test_catalogs_query_criteria _______________________________________________________

self = <astroquery.mast.tests.test_mast_remote.TestMast object at 0x14daf9d00>

    def test_catalogs_query_criteria(self):
    
        # clear columns config
        mast.Catalogs._column_configs = dict()
    
        # without position
        result = mast.Catalogs.query_criteria(catalog="Tic", Bmag=[30, 50], objType="STAR")
        assert isinstance(result, Table)
        assert len(result) >= 10
        assert result[np.where(result['ID'] == '81609218')]
    
        result = mast.Catalogs.query_criteria(catalog="ctl", Tmag=[10.5, 11], POSflag="2mass")
        assert isinstance(result, Table)
>       assert len(result) >= 400
E       assert 316 >= 400
E        +  where 316 = len(<Table masked=True length=316>\n    ID    version   HIP      TYC      ... e_Dec_orig raddflag wdflag   objID  \n   str9 ......       70.0        1      0 137129648\n394245807 20190415    -- 9490-00757-1 ...      130.0        1      0 516829248)

astroquery/mast/tests/test_mast_remote.py:463: AssertionError
```

File: astroquery/mast/cutouts.py (hunk 2 above)
Comment (jaymedina) [reply astroquery/mast/cutouts.py:135]: Hi @bsipocz , I've been tasked with addressing threads for this PR so it can get merged: From what I can tell it looks like there are two separate databases, and setting `moving_target` will make the request call for the moving target database in MAST.

File: astroquery/mast/observations.py (hunk 3 above)
Comment (jaymedina) [reply astroquery/mast/observations.py:549]: This looks like an unrelated commit that was part of something else. I can make another commit to reverse these changes on my local branch.

File: astroquery/mast/cutouts.py (hunk 2 above)
Comment (ceb8) [reply astroquery/mast/cutouts.py:135]: Hi, so my reason for doing this was twofold:
- The resolving method for moving vs regular targets is completely different, and naming in astronomy is not always unique so I was worried about the user putting in a valid moving target and getting a stationary result.
- The output for moving targets is not quite the same as for regular targets, since it's a different type of cutout, so I thought it was clearer to keep the two queries more separate.

Both of these concerns could be addressed by having a moving target flag that defaults to false instead of a whole new argument.

File: astroquery/mast/cutouts.py (hunk 1 above)
Comment (ceb8) [reply astroquery/mast/cutouts.py:168]: We rec would be to file and issue and have this be a seperate PR, bc this is the convention throughout the MAST module, so I think it would be clearer if they all got changed at once.

File: docs/mast/mast.rst (hunk 6 above)
Comment (ceb8) [reply docs/mast/mast.rst:820]: @jaymedina Try this example now, if I remember correctly the JPL horizons service was having timeout problems last August, so that might have been the problem.

Comment (jaymedina) [reply docs/mast/mast.rst:820]: Looks good

```
In [3]: hdulist = Tesscut.get_cutouts(moving_target="Eleonora", size=5, sector=6
   ...: )

In [4]: hdulist[0].info()
Filename: <class '_io.BytesIO'>
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU      54   ()      
  1  PIXELS        1 BinTableHDU    150   355R x 16C   [D, E, J, 25J, 25E, 25E, 25E, 25E, J, E, E, 38A, D, D, D, D]   
  2  APERTURE      1 ImageHDU        97   (2136, 2078)   int32   
  ```

Comment (jaymedina) [reply docs/mast/mast.rst:820]: Looks like we have 355 rows now instead of 356. Must be a numpy thing?

File: docs/mast/mast.rst (hunk 4 above)
Comment (jaymedina) [reply docs/mast/mast.rst:298]: Just checked the rest of `mast.rst` and the indentation is consistent with the rest of the document. I worry pulling it back will mess up the formatting.

File: docs/mast/mast.rst (hunk 6 above)
Comment (jaymedina) [reply docs/mast/mast.rst:820]: @ceb8 would you mind giving me writing access to your forked astroquery repo? I have some updates I need to commit to this branch.

Comment (ceb8) [reply docs/mast/mast.rst:820]: @jaymedina you should have an invite now

**Comment by jaymedina** [comment]
I'll have to dig myself out of this again. In the meantime, tests look good. There's one test that was intentionally skipped because it retrieves too much data, which I'll revisit later - it's lower priority because I have some back-end tests running and this test executes, I'll take a look at it in the next sprint and see if I can recreate it for the GitHub tests: 
```
test_mast_remote.py ..................s....................                                                                                                                                    [100%]

====================================================================================== short test summary info =======================================================================================
SKIPPED [1] test_mast_remote.py:260: Tests should not download this much data. Skipping until revised.
============================================================================= 38 passed, 1 skipped in 1521.35s (0:25:21) =============================================================================
```

**Comment by jaymedina** [comment]
I'm assuming a rebase will collapse all the commits that got merged from recent changes to `main` @bsipocz @ceb8

**Comment by ceb8** [comment]
@jaymedina Yes, a rebase is definitely in order, if you do that and are still having issued with stray commits, let me know. We might have to back out your changes, then rebase, then add them back in.

File: astroquery/mast/observations.py (hunk 3 above)
Comment (jaymedina) [reply astroquery/mast/observations.py:549]: I'm going to revert these changes to what's in the `main` branch because it's following the same logic just written slightly differently, without the `fall_back` trigger and a few other things.

File: astroquery/mast/cutouts.py [hunk 7]
```diff
@@ -104,10 +104,13 @@ def __init__(self):
...
         services = {"sector": {"path": "sector"},
-                    "astrocut": {"path": "astrocut"}}
+                    "astrocut": {"path": "astrocut"},
+                    "mt_sector": {"path": "moving_target/sector"},
+                    "mt_astrocut": {"path": "moving_target/astrocut"}
+                    }
         self._service_api_connection.set_service_params(services, "tesscut")
 
-    def get_sectors(self, coordinates=None, radius=0*u.deg, objectname=None):
+    def get_sectors(self, coordinates=None, radius=0*u.deg, objectname=None, moving_target=None, mt_type=None):
```
[the remark this answers is not in the document]
Comment (jaymedina) [reply astroquery/mast/cutouts.py:113]: Whatever gets included in `*` also needs to be added to `params = {"obj_id": moving_target}` in L157 before the service call right?

[the remark this answers is not in the document]
Comment (bsipocz) [reply astroquery/mast/cutouts.py:113]: `*` means that the rest of the parameters are keyword only, nothing additional is included in it.

File: astroquery/mast/cutouts.py (hunk 2 above)
Comment (bsipocz) [reply astroquery/mast/cutouts.py:135]: The naming mess affects not just moving targets, I'm sure there is overlap within non-moving ones if we look closely, too, and also, the users don't need to know what's under the hood, that depending on the type of the object the query goes to a different database. 
So, I find the logic of having a bool `moving_target` clearer than basically having two separate APIs for the name depending on the service.

File: docs/mast/mast.rst (hunk 4 above)
Comment (bsipocz) [reply docs/mast/mast.rst:298]: OK, then this is an issue for a separate quick cleanup PR.

File: astroquery/mast/cutouts.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jaymedina) [reply astroquery/mast/cutouts.py:113]: Oh I see - learned something new today. This might be something worth starting a new PR over, since its a syntax change, we'd want to add the `*` to all the Zcut methods as well, etc. What do you think?

File: astroquery/mast/cutouts.py (hunk 2 above)
Comment (jaymedina) [reply astroquery/mast/cutouts.py]: I've made a new commit that turns `moving_target` into an optional bool argument. With your approvals, I'll make another commit to update the docstrings, tests, documentation, and collapse the commits into one.

**Comment by jaymedina** [comment]
Heads up @ceb8 & @bsipocz : I'm going to move forward with those changes I made in commit 6e91621 unless there are any objections.

**Comment by jaymedina** [comment]
Since `moving_target` will now be a bool value, `moving_target` and `objectname` will now be required to go together as specified in the dosctrings. Tests will account for the possible user input combinations below:

1. moving_target only 
2. coordinates only (correct)
3. objectname only 
4. moving_target + coordinates
5. moving_target + objectname (correct) 
6. coordinates + objectname
7. moving_target + coordinates + objectname

**Comment by jaymedina** [comment]
The old dependencies and readthedocs checks are failing from not having the recent merge. Is it alright if I rebase really quick @bsipocz ? I have a backup branch ready.

File: astroquery/mast/cutouts.py [hunk 8]
```diff
@@ -164,7 +195,8 @@ def get_sectors(self, coordinates=None, radius=0*u.deg, objectname=None):
             warnings.warn("Coordinates are not in any TESS sector.", NoResultsWarning)
         return Table(sector_dict)
 
-    def download_cutouts(self, coordinates=None, size=5, sector=None, path=".", inflate=True, objectname=None):
+    def download_cutouts(self, coordinates=None, size=5, sector=None, path=".", inflate=True,
```
[the remark this answers is not in the document]
Comment (jaymedina) [reply astroquery/mast/cutouts.py:207]: I've made a separate issue for this: https://github.com/astropy/astroquery/issues/2267

**Comment by bsipocz** [comment]
> The old dependencies and readthedocs checks are failing from not having the recent merge. Is it alright if I rebase really quick @bsipocz ? I have a backup branch ready.

No need to ask, do feel free to rebase, squash commits, etc as often as you need :)

**Comment by ceb8** [comment]
@jaymedina  This is looking good.  Is it ready to be merged, or are you planning to do anything else?

**Comment by jaymedina** [comment]
> @jaymedina This is looking good. Is it ready to be merged, or are you planning to do anything else?

Nope I’m done with the PR. Now just trying to figure out why the readthedocs check is failing, but the error messages aren’t giving me any hints.

**Comment by ceb8** [comment]
@jaymedina OK, I'll give it a last look over, and see whats up with the docs check in the next day or so.

File: docs/mast/mast.rst [hunk 9]
```diff
@@ -799,13 +812,33 @@ ccd a target pixel file will be produced for each one.
...
 
 
-The `~astroquery.mast.TesscutClass.download_cutouts` function takes a coordinate or object name
-(such as "M104" or "TIC 32449963") and cutout size (in pixels or an angular quantity) and
-downloads the cutout target pixel file(s).
+Requesting a cutout by moving_target accesses the
+`MAST Moving Target TESScut API <https://mast.stsci.edu/tesscut/docs/getting_started.html#moving-target-cutouts>`__
+and returns a target pixel file, with format described
+`here <https://astrocut.readthedocs.io/en/latest/astrocut/file_formats.html#path-focused-target-pixel-files>`__.
+The moving_target is an optional bool argument where `True` signifies that the accompanying `objectname` input is the object name or ID understood by the
```
Comment (eerovaher) [inline docs/mast/mast.rst:819]: The raw log files of the documentation builds are available from the URL `https://readthedocs.org/api/v2/build/{id}.txt` where `{id}` is an index that gets incremented every time a new build in started. Once the index values for the last successful and first failing build have been looked up it is possible to use the following shell function to quickly see what caused the failure:
```shell
check_readthedocs() {
  diff <(curl -s https://readthedocs.org/api/v2/build/$1.txt | sort) <(curl -s https://readthedocs.org/api/v2/build/$2.txt | sort) | grep WARNING;
}
```
In this case the relevant build indices are 15919123 and 15919240, and the output of `check_readthedocs 15919123 15919240` is
```
> /home/docs/checkouts/readthedocs.org/user_builds/astroquery/checkouts/2121/docs/mast/mast.rst:815: WARNING: py:obj reference target not found: objectname
```
This points out the problem and its solution quite clearly:
```suggestion
The moving_target is an optional bool argument where `True` signifies that the accompanying ``objectname`` input is the object name or ID understood by the
```

**Comment by ceb8** [comment]
Thanks @jaymedina !

**Comment by jaymedina** [comment]
Thank you :) @ceb8

**Comment by bsipocz** [comment]
This needs a changelog entry, please open a separate PR for adding one, with using the number of this PR.
